File: src/main/java/komu/blunt/parser/SourceReader.java
package komu.blunt.parser;

import static com.google.common.base.Preconditions.checkNotNull;

final class SourceReader {

    private final String source;
    private int line = 1;
    private int column = 1;
    private int position = 0;

    public SourceReader(String source) {
        this.source = checkNotNull(source);
    }

    public int read() {
        if (!hasMore()) return -1;

        char ch = source.charAt(position++);
        if (ch == '\n') {
            line++;
            column = 1;
        } else {
            column++;
        }
        
        return ch;
    }
    
    public boolean matches(String s) {
        assert !s.isEmpty();
        
        if (!hasMore()) return false;
        
        return source.regionMatches(position, s, 0, s.length());
    }

    public int getLine() {
        return line;
    }

    public int getColumn() {
        return column;
    }

    public int peek() {
        if (hasMore())
            return source.charAt(position);
        else
            return -1;
    }

    private boolean hasMore() {
        return position < source.length();
    }

    public SourceLocation getLocation() {
        return new SourceLocation(line, column);
    }
}

File: src/main/kotlin/komu/blunt/parser/SourceReader.kt
package komu.blunt.parser

class SourceReader(private val source: String) {

    private var _line = 1
    private var _column = 1
    private var _position = 0

    fun read(): Char? {
        if (!hasMore) return null

        val ch = source[_position++]
        if (ch == '\n') {
            _line++
            _column = 1
        } else {
            _column++
        }

        return ch
    }

    fun matches(s: String): Boolean {
        if (s == "") throw IllegalArgumentException("empty string")

        if (!hasMore) return false

        return source.regionMatches(position, s, 0, s.length())
    }

    fun peek(): Char? =
        if (hasMore)
            source[_position]
        else
            null

    val hasMore: Boolean
        get() = _position < source.length()

    val line: Int
        get() = _line

    val column: Int
        get() = _column

    val location: SourceLocation
        get() = SourceLocation(_line, _column)

    var position: Int
        get() = _position
        set(position) {
            check(position >= 0 && position < source.length)
            this._position = position
        }

    fun save() = SourceReaderState(line, column, position)

    fun restore(state: SourceReaderState) {
        _line = state.line
        _column = state.column
        _position = state.position
    }
}

class SourceReaderState (val line: Int, val column: Int, val position: Int)
